src/Microsoft.AspNet.OData.Shared/Query/ExpandQueryBuilder.cs

```diff
@@ -0,0 +1,121 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="ExpandQueryBuilder.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved. 
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Formatter;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Exposes the ability to generate a $expand query string from a payload object.
+    /// </summary>
+    public class ExpandQueryBuilder : IExpandQueryBuilder
+    {
+        /// <inheritdoc />
+        public virtual string GenerateExpandQueryString(object value, IEdmModel model)
+        {
+            if (value == null)
+            {
+                throw Error.ArgumentNull(nameof(value));
+            }
+
+            if (model == null)
+            {
+                throw Error.ArgumentNull(nameof(model));
+            }
+
+            return GenerateExpandQueryStringInternal(value, model, true);
+        }
+
+        private string GenerateExpandQueryStringInternal(object value, IEdmModel model, bool isFirstExpand)
+        {
+            Type type = value.GetType();
+            Type _elementType = null;
+            bool isCollection = TypeHelper.IsCollection(type, out _elementType);
+            IList objList = null;
+            if (isCollection)
+            {
+                type = _elementType;
+                objList = value as IList;
+            }
+
+            string edmFullName = type.EdmFullName();
+            IEdmSchemaType schemaType = model.FindType(edmFullName);
+            IEdmStructuredType edmStructuredType = schemaType as IEdmStructuredType;
+
+            IEnumerable<IEdmNavigationProperty> navigationProperties = edmStructuredType.NavigationProperties();
+            string expandString = "";
+
+            int count = 0;
+
+            List<string> navPropNames = new List<string>();
+
+            foreach (IEdmNavigationProperty navProp in navigationProperties)
+            {
+                count++;
+                PropertyInfo prop = type.GetProperty(navProp.Name);
+
+                if (isCollection)
+                {
+                    foreach (object obj in objList)
```

Comment:
@KenitoInc Maybe I'm having a problem following the code or looping through the items in the collection is redundant.
Consider that your collection is:
```csharp
var orders = new List<Order>
{
    new Order { Id = 1, Customer = new Customer { Id = 7 } },
    new Order { Id = 2, Customer = new Customer { Id = 13 } }
};
```
and `navProp` is `Customer`. Would the loop yield a different result against order 1 as opposed to order 2? Does it not suffice to process the first item as a representative of the items in the collection?

Reply:
The problem comes when Order 2 has another nested navigation property which is not present in Order 1

Reply:
@KenitoInc You mean like this?
```
var orders = new List<Order>
{
    new Order { Id = 1 },
    new Order { Id = 2, Customer = new Customer { Id = 13 } }
};
```
Interesting...